lib/src/types/transaction/instruction.dart

```diff
@@ -0,0 +1,78 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:solana/src/types/transaction/transaction.dart';
+
+part 'instruction.freezed.dart';
+part 'instruction.g.dart';
+
+const _transferParsedInstruction = 'transfer';
+
+/// An instruction which is part of a [TxMessage]
+@freezed
+class TxInstruction {
```

Comment:
Freezed is capable of decoding union cases, so it can be done like this:

```dart
/// An instruction which is part of a [TxMessage]
@Freezed(unionKey: 'program', fallbackUnion: 'unsupported')
class TxInstruction with _$TxInstruction {
  const factory TxInstruction.system({
    required String programId,
    required ParsedInstruction parsed,
  }) = TxInstructionSystem;

  /// Instruction representing a memo with content [memo]
  @FreezedUnionValue('spl-memo')
  const factory TxInstruction.memo({
    required String parsed,
  }) = TxInstructionMemo;

  const factory TxInstruction.unsupported({
    required String program,
  }) = TxInstructionUnsupported;

  factory TxInstruction.fromJson(Map<String, dynamic> json) =>
      _$TxInstructionFromJson(json);
}

@Freezed(unionKey: 'type', fallbackUnion: 'unsupported')
class ParsedInstruction with _$ParsedInstruction {
  const factory ParsedInstruction.transfer({
    required TransferInfo info,
  }) = ParsedInstructionTransfer;

  /// Instruction representing any unsupported instructions
  /// that might be found in a transaction
  const factory ParsedInstruction.unsupported({
    required String type,
  }) = ParsedInstructionUnsupported;

  factory ParsedInstruction.fromJson(Map<String, dynamic> json) =>
      _$ParsedInstructionFromJson(json);
}

@freezed
class TransferInfo with _$TransferInfo {
  const factory TransferInfo({
    required int lamports,
    required String source,
    required String destination,
  }) = _TransferInfo;

  factory TransferInfo.fromJson(Map<String, dynamic> json) =>
      _$TransferInfoFromJson(json);
}
```